### DISCUSSION TITLE
Code coverage during integration testing

### DISCUSSION THREAD

**Discussion by ThibFrgsGmz** [nasa/fprime#1530] [Q&A]
Hi,

When running integration tests, using [Integration Test API](https://github.com/nasa/fprime/blob/devel/docs/UsersGuide/dev/testAPI/user_guide.md#gds-integration-test-api-user-guide), is it possible to know the code coverage of the deployment?

Depending on the project, we have to determine the code coverage level of a deployment during the system tests.

**Comment by Joshua-Anderson** [comment]
It may not quite be possible out of the box, but we already have code coverage support for unit testing, so all that's needed is the ability to compile a deployment executable with the same code coverage flags.

**Reply by ThibFrgsGmz** [reply]
So I have to generate a build cache by passing some options? Like `fprime-util generate -some_flag=get_coverage`

**Reply by Joshua-Anderson** [reply]
Yeah, you'll want to set the same flags coverage does, then start your ref binary, run the tests, and use gcovr to examine coverage.

```
FPRIME_TESTING_REQUIRED_COMPILE_FLAGS -fprofile-arcs -ftest-coverage
FPRIME_TESTING_REQUIRED_LINK_FLAGS --coverage
```

**Reply by ThibFrgsGmz** [reply]
OK, I see where they need to be defined: 

https://github.com/nasa/fprime/blob/7bf79db61b1b63d60a624328dc909ee62869e8ba/cmake/target/build.cmake#L17
https://github.com/nasa/fprime/blob/7bf79db61b1b63d60a624328dc909ee62869e8ba/cmake/target/build.cmake#L18

To do something clean, I would have to create and activate a `FPRIME_ENABLE_INTEGRATION_COVERAGE` symbol?

So theoretically I add these coverage flags when creating my deployment: `fprime-util generate` `fprime-util build -j32`

Then I launch the deployment `./Ref` and the backend of GDS: `fprime-gds -n -g none`.
I run the integration test suite with `pytest` and I expect to have somewhere the `Excel` test report ?

**Reply by LeStarch** [reply]
Perhaps I am a bit late to the discussion, but couldn't you do this:

1. Build the UT build:
```
cd Ref
fprime-util generate --ut
fprime-util build --ut
```

2. Then run the UT executable?  It has a few different flags compiled in, but is essentially the same binary.

```
fprime-gds -n ...
./build-fprime-automatic-native-ut/bin/Ref ...
pytest
```

3. Calculate coverage:
```
gcovr
```

**Reply by LeStarch** [reply]
This approach uses the UT build which already builds everything with the coverage flag.  You test like normal with the "coverage-enabled" binary.  Then calculate coverage.

We may choose to add a "no-check"  or "existing" option to `fprime-util check --coverage` to allow it to bypass running the UTs but keep all the hard-fought `gcovr` flags.

**Reply by ThibFrgsGmz** [reply]
Sorry for the delay, this study went on in the background! Thanks for the suggestion!

Unless I am mistaken, the command `fprime-util build --ut` does not generate a `./build-fprime-automatic-native-ut/` folder in which the binary of the whole software system and the topology dictionary are located.

In step 2, you open a terminal for each command and also for step 3? Where do you run `gcovr` from?
